Level-2/her/her.cpp

```diff
@@ -0,0 +1,159 @@
+
+    std::vector<hipFloatComplex> hAGold(sizeA); 
+
+    // initialize simple data for simple host side reference computation
+    helpers::matIdentity(hA.data(), N, N, lda);
+    hAGold = hA;
+
+    // using rocblas API
+    rocblas_handle handle;
+    rstatus = rocblas_create_handle(&handle);
+    CHECK_ROCBLAS_STATUS(rstatus);
+
+    {
+        // Naming: dX is in GPU (device) memory. hK is in CPU (host) memory
+
+        // allocate memory on device
+        helpers::DeviceVector<rocblas_float_complex> dA(sizeA);
+        helpers::DeviceVector<rocblas_float_complex> dX(sizeX);
+
+
+        if((!dA && sizeA) || (!dX && sizeX))
+        {
+            CHECK_HIP_ERROR(hipErrorOutOfMemory);
+            return EXIT_FAILURE;
+        }
+
+        // time data to device, computation, and data from device back to host
+        helpers::GPUTimer gpuTimer;
+        gpuTimer.start();
+
+        // copy data from CPU to device (all 3 complex types same memory layout)
+        CHECK_HIP_ERROR(hipMemcpy(dA, hA.data(), sizeof(hipFloatComplex) * sizeA, hipMemcpyHostToDevice));
+        CHECK_HIP_ERROR(hipMemcpy(dX, hX.data(), sizeof(std::complex<float>) * sizeX, hipMemcpyHostToDevice));
+
+        // enable passing alpha and beta parameters from pointer to host memory
+        rstatus = rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host);
+        CHECK_ROCBLAS_STATUS(rstatus);
+
+        // asynchronous calculation on device, returns before finished calculations
+        rstatus = rocblas_cher(handle, uplo, N, &hAlpha, dX, incx, dA, lda);
+
+        // check that calculation was launched correctly on device, not that result
+        // was computed yet
+        CHECK_ROCBLAS_STATUS(rstatus);
+
+        // fetch device memory results, automatically blocked until results ready
+        CHECK_HIP_ERROR(hipMemcpy(hA.data(), dA, sizeof(hipFloatComplex) * sizeA, hipMemcpyDeviceToHost));
+
+        gpuTimer.stop();
+
+    } // release device memory via helpers::DeviceVector destructors
+
+    std::cout << "alpha, N, lda = " << hAlpha << ", " << N << ", " << lda << std::endl;
+
+    // calculate expected result using CPU
+    for(int i = 0; i < N; i++)
+    {
+        // matrix is identity so just doing simpler calculation over x vectors
+
+        for (int j = 0; j < N; j++)
+        {
+            std::complex<float> r = hX[i] * std::conj(hX[j]);
+            r *= std::complex<float>(hAlpha, 0);
+
+            // using hip helper function hipCaddf to add hip complex type
+            hAGold[i*lda + j] = hipCaddf( hipFloatComplex(r.real(), r.imag()), hAGold[i*lda + j] );
+        }
+    }
+
+    bool fail = false;
+    for(int i = 0; i < N; i++)
+    {
+        for (int j = 0; j < N; j++)
+        {
+            if (uplo == rocblas_fill_upper && j > i) continue;
+            else if (uplo != rocblas_fill_upper && j < i) continue;
+
+            if (hAGold[i*lda+j] != hA[i*lda+j]) fail = true;
```

Comment:
Having an upper triangular matrix could result in more robust checking of the complex data type (i.e.), comparing the absolute value of complex numbers of `hAGold` and `hA` and normal comparison of diagonal elements of `hAGold` and `hA` as they would be real.

Reply:
As the diagonal imaginary is set to zero the equality can be used on and off diagonal, I just exclude the un-referenced portion of the GPU matrix results.    Not sure what you mean by robustness here, operator== of the hipFloatComplex is used here for exact match which requires complex mult and addition on GPU.    This is use of operator != code that I think may be dropped, but that is a future discussion.  I could flip the usage of std::complex and hipComplex for that...

Reply:
What I meant was, As per my understanding, if `hAGold` and `hA` are both complexes. Does it make sense to compare `hAGold` and `hA` by taking the absolute value of the complex numbers or compare the real and imaginary value directly as in this command here if `(hAGold[i*lda+j] != hA[i*lda+j]) fail = true;`?

Reply:
So by absolute value you would only be checking their vector length matches, so I am unclear why that would be preferred to exact match of CPU and GPU results?

Reply:
I usually compared two complex numbers by taking the modulus of it like this `if (abs(hAGold[i*lda+j]) != abs(hA[i*lda+j])) fail = true;` and I did not know comparing it directly `if (hAGold[i*lda+j] != hA[i*lda+j]) fail = true;` whether it will compare the real part or the imaginary part.

Reply:
The operator != and == are defined (for all 3 current variants of complex used in this example) which do as one would expect to require both components to match.  The modulus would if used for this purpose would incorrectly say all complex plane rotations are equal, e.g. 0+i1 should not match 1+0i.

Reply:
Ahh, you are right! Sorry for the confusion. If the operators `!=` and `==` compares both real and imaginary part that is great.